db/write_thread.cc

```diff
@@ -310,137 +356,295 @@ size_t WriteThreadImpl::EnterAsBatchGroupLeader(
-  w->sequence = sequence;
+  leader->write_group = write_group;
+  write_group->leader = leader;
+  write_group->size = 1;
+  Writer* last_writer = leader;
 
-  // Initialize and wake up the others
-  while (w != pg->last_writer) {
-    // Writers that won't write don't get sequence allotment
-    if (!w->CallbackFailed() && w->ShouldWriteToMemtable()) {
-      // There is a sequence number of each written key
-      sequence += WriteBatchInternal::Count(w->batch);
+  if (!allow_concurrent_memtable_write_ || !leader->batch->HasMerge()) {
+    Writer* newest_writer = newest_memtable_writer_.load();
+    CreateMissingNewerLinks(newest_writer);
+
+    Writer* w = leader;
+    while (w != newest_writer) {
+      w = w->link_newer;
+
+      if (w->batch == nullptr) {
+        break;
+      }
+
+      if (w->batch->HasMerge()) {
+        break;
+      }
+
+      if (!allow_concurrent_memtable_write_) {
+        auto batch_size = WriteBatchInternal::ByteSize(w->batch);
+        if (size + batch_size > max_size) {
+          // Do not make batch too big
+          break;
+        }
+        size += batch_size;
+      }
+
+      w->write_group = write_group;
+      last_writer = w;
+      write_group->size++;
     }
-    w = w->link_newer;
+  }
+
+  write_group->last_writer = last_writer;
+  write_group->last_sequence =
+      last_writer->sequence + WriteBatchInternal::Count(last_writer->batch) - 1;
+}
 
-    w->sequence = sequence;  // sequence number for the first key in the batch
-    w->parallel_group = pg;
-    SetState(w, STATE_PARALLEL_FOLLOWER);
+void WriteThread::ExitAsMemTableWriter(Writer* self, WriteGroup& write_group) {
+  Writer* leader = write_group.leader;
+  Writer* last_writer = write_group.last_writer;
+
+  Writer* newest_writer = last_writer;
+  if (!newest_memtable_writer_.compare_exchange_strong(newest_writer,
+                                                       nullptr)) {
+    CreateMissingNewerLinks(newest_writer);
+    Writer* next_leader = last_writer->link_newer;
+    assert(next_leader != nullptr);
+    next_leader->link_older = nullptr;
+    SetState(next_leader, STATE_MEMTABLE_WRITER_LEADER);
+  }
+  auto v = write_group.ToVector();
+  for (Writer* w : v) {
+    if (!write_group.status.ok()) {
+      w->status = write_group.status;
+    }
+    // Leader has to wait for the operation finish, because it owns the write
+    // group.
+    if (w != leader) {
+      SetState(w, STATE_COMPLETED);
+    }
+  }
+  SetState(leader, STATE_COMPLETED);
+}
+
+void WriteThread::LaunchParallelMemTableWriters(WriteGroup& write_group) {
+  // EnterAsBatchGroupLeader already created the links from leader to
+  // newer writers in the group
+  // Initialize and wake up the others
+  size_t num_writers = 0;
+  for (auto w : write_group) {
+    if (w->ShouldWriteToMemtable()) {
+      num_writers++;
+    }
+  }
+  write_group.running.store(num_writers);
+  for (auto w : write_group) {
+    if (w->ShouldWriteToMemtable()) {
```

Comment:
The will be set to completed state in ExitAsMemTableWriter(). I changed it back to launching everyone in the group for simplicity in #2286